packages/contentful-slatejs-adapter/src/slatejs-to-contentful-adapter.ts

```diff
@@ -1,70 +1,80 @@
 import flatMap from 'lodash.flatmap';
-import * as Contentful from '@contentful/rich-text-types';
-import { ContentfulNode, SlateNode } from './types';
-import { getDataOfDefault } from './helpers';
+import { getDataOrDefault } from './helpers';
 import { SchemaJSON, Schema, fromJSON } from './schema';
 
+import * as Contentful from '@contentful/rich-text-types';
+import {
+  ContentfulNode,
+  ContentfulElementNode,
+  SlateNode,
+  SlateElement,
+  SlateText,
+  SlateMarks,
+} from './types';
+
 export interface ToContentfulDocumentProperties {
-  document: Slate.Document;
+  document: SlateNode[];
   schema?: SchemaJSON;
 }
 
 export default function toContentfulDocument({
   document,
   schema,
 }: ToContentfulDocumentProperties): Contentful.Document {
+  // TODO:
+  // We allow adding data to the root document node, but Slate >v0.5.0
+  // has no concept of a root document node. We should determine whether
+  // this will be a compatibility problem for existing users.
   return {
     nodeType: Contentful.BLOCKS.DOCUMENT,
-    data: getDataOfDefault(document.data),
+    data: {},
     content: flatMap(
-      document.nodes,
+      document,
       node => convertNode(node, fromJSON(schema)) as Contentful.Block[],
     ),
   };
 }
 
-function convertNode(node: SlateNode, schema: Schema): ContentfulNode[] {
+function convertNode(
+  node: SlateNode,
+  schema: Schema
+): ContentfulNode[] {
   const nodes: ContentfulNode[] = [];
-  switch (node.object) {
-    case 'block':
-    case 'inline':
-      const slateNode = node as Slate.Block;
-      const content = flatMap(slateNode.nodes, childNode => convertNode(childNode, schema));
-      if (!slateNode.type) {
-        throw new Error(`Unexpected slate node ${JSON.stringify(slateNode)}`);
-      }
-
-      const contentfulBlock: Contentful.Block = {
-        nodeType: slateNode.type,
-        content: [],
-        data: getDataOfDefault(slateNode.data),
-      };
-
-      if (!schema.isVoid(contentfulBlock)) {
-        contentfulBlock.content = content;
-      }
-      nodes.push(contentfulBlock);
-      break;
-    case 'text':
-      convertText(node as Slate.Text).forEach(childNode => nodes.push(childNode));
-      break;
-    default:
-      assertUnreachable(node);
-      break;
+  if (isSlateElement(node)) {
+    const contentfulElement: ContentfulElementNode = {
+      nodeType: node.type,
+      data: getDataOrDefault(node.data),
+      content: [],
+    };
+    if (!schema.isVoid(contentfulElement)) {
+      contentfulElement.content = flatMap(node.children, childNode => convertNode(childNode, schema));
+    }
+    nodes.push(contentfulElement);
+  } else {
+    const contentfulText = convertText(node);
+    nodes.push(contentfulText);
   }
-
   return nodes;
 }
 
-function convertText(node: Slate.Text): Contentful.Text[] {
-  return node.leaves.map<Contentful.Text>(leaf => ({
+function convertText(node: SlateText): Contentful.Text {
```

Comment:
Just out of curiosity, how does `SlateText` differ from a `Contentful.Text`? What's the main difference between them?

Reply:
A slate text node looks like

```json
{
  "text": "My cool text string",
  "data": {},
  "bold": true
}
```

A contentful text node looks like this

```js
{
  "nodeType": "text",
  "value": "My cool text string",
  "data": {},
  "marks": [
    { "type": "bold" }
  ]
}
```